### PR TITLE
[Pipeline/Tooling] Improve change detection for running system-probe kitchen tests

### PR DESCRIPTION
**Pull request by Pythyu**
### What does this PR do?

It improve kitchen system-probe file change detection by checking manually if any file got changed with a new job ``get_file_diff`` that gets run at the start.

### Motivation

Previously we used the gitlab ``changes:`` functionality to check for any changes in files. This feature wasn't working on the first commit on a new branch. Meaning that every new branch had system-probe kitchen tests running on the first commit.

### Additional Notes

On the ``get_file_diff`` job we get the diff from the last common commit with the current branch and the base_branch, and the HEAD. This output is multiline but dotenv format doesn't support multiline variable (at least from my tests) so I had to make them into a one liner that is separated by litteral "\n" (it could have been any seperator)

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: .gitlab/functional_test/system_probe.yml [hunk 1]
```diff
@@ -11,17 +11,26 @@
...
+    job: get_file_diff
+    artifacts: true
   stage: functional_test
   variables:
     AGENT_MAJOR_VERSION: 7
     DD_PIPELINE_ID: $CI_PIPELINE_ID-fnct
     CHEF_VERSION: 14.15.6
   script:
+    - set +e
+    - echo "$DIFF_FILES" | grep "pkg/collector/corechecks/ebpf/" || echo "$DIFF_FILES" | grep "pkg/ebf/" || echo "$DIFF_FILES" | grep "pkg/network/" || echo "$DIFF_FILES" | grep "pkg/util/kernel" || echo "$DIFF_FILES" | grep "test/kitchen/site-cookbooks/dd-system-probe-check/" || echo "$DIFF_FILES" | grep "test/kitchen/test/integration/system-probe-test/" || echo "$DIFF_FILES" | grep "test/kitchen/test/integration/win-sysprobe-test/" || echo "$DIFF_FILES" | grep ".gitlab/functional_test/system_probe.yml" || echo "$DIFF_FILES" | grep "test/new-e2e/system-probe/" || echo "$DIFF_FILES" | grep "test/new-e2e/scenarios/system-probe/"; export exit_code=$?
```
Comment (hmahmood) [inline .gitlab/functional_test/system_probe.yml:24]: I think this would be more readable if you put the path match list in a variable and this whole bit for matching into a separate script that you can reference from multiple job definitions. You can loop over the variable in the script instead of listing out each individual grep.

Comment (Pythyu) [reply .gitlab/functional_test/system_probe.yml:24]: I agree it's a way more cleaner way. I'm going to try that 👍

**Review by sgnn7 — APPROVED** [review]
LGTM for ASC-owned `release.json`

File: .gitlab/functional_test/system_probe.yml (hunk 1 above)
Comment (Pythyu) [reply .gitlab/functional_test/system_probe.yml]: Should be better now :)

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -139,6 +139,17 @@ variables:
   S3_DSD6_URI: s3://dsd6-staging
   RELEASE_VERSION_6: nightly
   RELEASE_VERSION_7: nightly-a7
+  SYSTEM_PROBE_SRC_PATH: |
```
Comment (hmahmood) [inline .gitlab-ci.yml:142]: I would put this under `.on_system_probe_changes_or_manual`

File: .gitlab/functional_test/system_probe.yml [hunk 3]
```diff
@@ -11,17 +11,36 @@
...
+    job: get_file_diff
+    artifacts: true
   stage: functional_test
   variables:
     AGENT_MAJOR_VERSION: 7
     DD_PIPELINE_ID: $CI_PIPELINE_ID-fnct
     CHEF_VERSION: 14.15.6
   script:
+    - set +e
+    - export SYSTEM_PROBE_PATHS=$(echo "$SYSTEM_PROBE_SRC_PATH")
```
Comment (hmahmood) [inline .gitlab/functional_test/system_probe.yml:24]: Can we put this under `.on_system_probe_changes_or_manual` and reference it here and other places using [!reference](https://docs.gitlab.com/ee/ci/yaml/yaml_optimization.html#reference-tags) tags?

Comment (hmahmood) [reply .gitlab/functional_test/system_probe.yml:24]: Also was `$SYSTEM_PROBE_SRC_PATH` not available as is here?

File: .gitlab/kernel_version_testing/system_probe.yml [hunk 4]
```diff
@@ -1,24 +1,58 @@
...
+get_file_diff:
+  image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-buildimages/deb_x64$DATADOG_AGENT_BUILDIMAGES_SUFFIX:$DATADOG_AGENT_BUILDIMAGES
+  tags: ["runner:main"]
+  rules:
+    !reference [ .on_system_probe_changes_or_manual ]
+  needs: []
+  stage: kernel_matrix_testing
+  script:
+    - git fetch origin main:main
+    - echo "DIFF_FILES=$(first=0;for element in $(git diff --name-only $(git merge-base $(inv release.get-release-json-value base_branch) HEAD) HEAD); do if [ $first -eq 0 ]; then first=1; else echo -n '\n'; fi; echo $element; done)" | tr -d '\n' > file_diff.env
```
Comment (brycekahle) [inline .gitlab/kernel_version_testing/system_probe.yml:50]: Could this be made an invoke task? I find this hard to follow.

**Review by brycekahle — CHANGES_REQUESTED** [review]

File: .gitlab/functional_test/system_probe.yml (hunk 3 above)
Comment (Pythyu) [reply .gitlab/functional_test/system_probe.yml:24]: I had some issues before with multiline gitlab variable to shell for loop, but I'll try again to be sure. And yea using a reference is good idea doing that o7

File: .gitlab/functional_test/system_probe.yml [hunk 5]
```diff
@@ -11,17 +11,36 @@
...
+    - |+
+      for path in $SYSTEM_PROBE_PATHS
```
[the remark this answers is not in the document]
Comment (Pythyu) [reply .gitlab/functional_test/system_probe.yml]: Had some previous issues with multiline gitlab variables but here it is indeed not needed. Fixed 👍

File: .gitlab/kernel_version_testing/system_probe.yml (hunk 4 above)
Comment (Pythyu) [reply .gitlab/kernel_version_testing/system_probe.yml]: Done 👍

File: .gitlab/functional_test/system_probe.yml (hunk 3 above)
Comment (Pythyu) [reply .gitlab/functional_test/system_probe.yml]: Done 👍

File: .gitlab-ci.yml [hunk 6]
```diff
@@ -824,21 +823,23 @@ workflow:
...
-      - test/kitchen/test/integration/win-sysprobe-test/**/*
-      - .gitlab/functional_test/system_probe.yml
-      - test/new-e2e/system-probe/**/*
-      - test/new-e2e/scenarios/system-probe/**/*
-    when: on_success
-  - when: manual
+  - variables:
+      SYSTEM_PROBE_SRC_PATH: |
+        pkg/collector/corechecks/ebpf/
+        pkg/ebf/
```
Comment (brycekahle) [inline .gitlab-ci.yml:829]:
```suggestion
        pkg/ebpf/
```

File: .gitlab-ci.yml [hunk 7]
```diff
@@ -824,21 +823,23 @@ workflow:
...
+        pkg/network/
+        pkg/util/kernel
```
Comment (brycekahle) [inline .gitlab-ci.yml:831]:
```suggestion
        pkg/util/kernel/
```

File: tasks/system_probe.py [hunk 8]
```diff
@@ -1501,3 +1501,16 @@ def test_microvms(
     ctx.run(
         f"cd ./test/new-e2e && go run ./scenarios/system-probe/main.go --name usama-saqib-test {go_args} --shutdown-period 720",
     )
+
+
+@task
+def generate_git_file_diff(ctx):
+    file_diff = ctx.run(
+        "git diff --name-only $(git merge-base $(inv release.get-release-json-value base_branch) HEAD) HEAD", hide=True
```
Comment (brycekahle) [inline tasks/system_probe.py:1509]: You can call `inv release.get-release-json-value base_branch` directly as a function rather than using a shell.

File: tasks/system_probe.py [hunk 9]
```diff
@@ -1501,3 +1501,16 @@ def test_microvms(
...
+    )
+    file_diff_lines = file_diff.stdout.split("\n")
+    for k, line in enumerate(file_diff_lines):
+        if k != 0:
+            print("\\n", end="")
+        print(line, end="")
```
Comment (brycekahle) [inline tasks/system_probe.py:1515]: Is this just re-adding the newlines that were removed with the `split("\n")`?

**Review by brycekahle — CHANGES_REQUESTED** [review]

File: tasks/system_probe.py (hunk 9 above)
Comment (Pythyu) [reply tasks/system_probe.py:1515]: No exactly it's replacing ``\n`` by `\\n`` but I was doing something else at the same time didn't I could just've used replace direclty. Fixed :+1

File: tasks/system_probe.py (hunk 8 above)
Comment (Pythyu) [reply tasks/system_probe.py:1509]: I can't because ``get_release_json_value`` print his output directly and capturing the python function output is equivalent to running it as is in shell. ref :
```python

@task(help={'key': "Path to the release.json key, separated with double colons, eg. 'last_stable::6'"})
def get_release_json_value(_, key):
    release_json = _load_release_json()

    path = key.split('::')

    for element in path:
        if element not in release_json:
            raise Exit(code=1, message=f"Couldn't find '{key}' in release.json")

        release_json = release_json.get(element)

    print(release_json)
```

Comment (brycekahle) [reply tasks/system_probe.py:1543]: That makes sense. I think you could extract a common function that could be called, but that can wait for a future PR.

File: tasks/system_probe.py [hunk 10]
```diff
@@ -1535,3 +1535,12 @@ def test_microvms(
...
         f"cd ./test/new-e2e && go run ./scenarios/system-probe/main.go --name usama-saqib-test {go_args} --shutdown-period 720",
     )
+
+
+@task
+def generate_git_file_diff(ctx):
+    file_diff = ctx.run(
+        "git diff --name-only $(git merge-base $(inv release.get-release-json-value base_branch) HEAD) HEAD", hide=True
+    )
+    file_diff_lines = file_diff.stdout.replace("\n", "\\n")
```
Comment (brycekahle) [inline tasks/system_probe.py:1545]: Why do you need to escape the newlines?

**Review by brycekahle — APPROVED** [review]
Minor remaining suggestion

File: tasks/system_probe.py (hunk 10 above)
Comment (Pythyu) [reply tasks/system_probe.py:1545]: I wrote about that in Additional Notes of the PR (dotenv doesn't handle multiline very well so a way to bypass that is to escape newline to make a one liner, once it get's send it will still get correctly iterated over by shell for loop). I'll add a comment about that

File: .gitlab/kernel_version_testing/system_probe.yml [hunk 11]
```diff
@@ -1,24 +1,61 @@
+---
+.system_probe_file_diff_check:
+  - set +e
+  - |+
+    for path in $SYSTEM_PROBE_SRC_PATH
+    do
+      echo "Checking $path..."
+      echo "$DIFF_FILES" | grep "$path"
+      export exit_code=$?
```
Comment (KSerrania) [inline .gitlab/kernel_version_testing/system_probe.yml:9]: Nit: for clarity, I'd give this variable a more specific name, like `file_diff_check_exit_code`. That way, it's easier to read the script of the jobs that use it, as:
```
- if [ ${file_diff_check_exit_code} -eq 0 ]; then
- <script>
- fi
```

File: tasks/system_probe.py [hunk 12]
```diff
@@ -1535,3 +1535,12 @@ def test_microvms(
     ctx.run(
         f"cd ./test/new-e2e && go run ./scenarios/system-probe/main.go --name usama-saqib-test {go_args} --shutdown-period 720",
     )
+
+
+@task
+def generate_git_file_diff(ctx):
```
Comment (KSerrania) [inline tasks/system_probe.py:1541]: This seems like a generic task that could be reused in other contexts, any reason why this should be put in the `system-probe` namespace?

**Review by KSerrania — COMMENTED** [review]
LGTM overall, left a few comments regarding positioning in the pipeline & naming.

File: tasks/system_probe.py (hunk 12 above)
Comment (Pythyu) [reply tasks/system_probe.py:1541]: No real reason apart from the fact that it's only used by system probe yet, I'll move it

File: .gitlab-ci.yml [hunk 13]
```diff
@@ -826,21 +825,23 @@ workflow:
     allow_failure: true
 
 .on_system_probe_changes_or_manual:
```
[the remark this answers is not in the document]
Comment (brycekahle) [reply .gitlab-ci.yml:827]: Can we retain the ability to run manually?

[the remark this answers is not in the document]
Comment (Pythyu) [reply .gitlab-ci.yml:827]: Since it will always run now, it will simply not execute if there's not change. So technically we can choose to re-run the job but it will not get executed because there's no change registered. So as it is we lost the abillity to run manually.
A current way if we want to keep the ability to run manually would be to have duplicate of the jobs that extend the same base but with a rule set as manual instead of the rule. What do you think about that way ?

**Comment by Pythyu** [comment]
I added duplicate of system probes jobs to allow manual start

File: .gitlab/functional_test/system_probe.yml [hunk 14]
```diff
@@ -143,5 +150,121 @@ kitchen_test_system_probe_windows_x64:
...
     - pushd $DD_AGENT_TESTING_DIR
     - tasks/kitchen_setup.sh
+    - !reference [ .system_probe_file_diff_check ]
+  script:
+    - tasks/run-test-kitchen.sh windows-sysprobe-test $AGENT_MAJOR_VERSION
+
+kitchen_test_system_probe_linux_x64_manual:
+  extends:
+    - .kitchen_test_system_probe
+    - .kitchen_azure_x64
```
Comment (brycekahle) [inline .gitlab/functional_test/system_probe.yml:160]: This is outdated now, we don't run the tests in Azure anymore, as of a couple days ago.

File: .gitlab/functional_test/system_probe.yml [hunk 15]
```diff
@@ -143,5 +150,121 @@ kitchen_test_system_probe_windows_x64:
     - export WINDOWS_DDNPM_SHASUM=$(inv release.get-release-json-value "$RELEASE_VERSION_7::WINDOWS_DDNPM_SHASUM")
     - pushd $DD_AGENT_TESTING_DIR
     - tasks/kitchen_setup.sh
+    - !reference [ .system_probe_file_diff_check ]
+  script:
+    - tasks/run-test-kitchen.sh windows-sysprobe-test $AGENT_MAJOR_VERSION
+
+kitchen_test_system_probe_linux_x64_manual:
```
Comment (brycekahle) [inline .gitlab/functional_test/system_probe.yml:157]: Is there a way to use a `!reference` to prevent copy/paste issues for the values that are exactly the same between the automatic and manual jobs?

Comment (hmahmood) [inline .gitlab/functional_test/system_probe.yml:157]: I am not sure having manual versions of all the jobs is a good idea. We should look at making a rule that will fallback to manual if there are no relevant file changes.

Comment (brycekahle) [reply .gitlab/functional_test/system_probe.yml:157]: It looks like the `CI_JOB_MANUAL` environment variable will be `true` if run manually, so we could skip the file check if so.
